Support for Vavr's Map keys & values validation

cardil opened the pull request:
This commit tests issue #2 - ignored tests will fail, as #2 is not supported.

alwins0n commented:
I have added the code to pass your tests. I will re-think if any cases are missing or more tests are needed. other than that it could be fine to merge.

what do you think @danieldietrich ?

cardil commented on src/main/java/io/vavr/beanvalidation2/valueextraction/MapKeyExtractor.java:
```diff
@@ -0,0 +1,35 @@
+ *  \  \/  /  /\  \  \/  /  /
+ *   \____/__/  \__\____/__/
+ *
+ * Copyright 2014-2017 Vavr, http://vavr.io
```
`2014-2019`

alwins0n replied:
copy pasta :( time to get my licence header right in the IDE

cardil commented on src/main/java/io/vavr/beanvalidation2/valueextraction/MapValueExtractor.java:
```diff
@@ -0,0 +1,35 @@
+ *  \  \/  /  /\  \  \/  /  /
+ *   \____/__/  \__\____/__/
+ *
+ * Copyright 2014-2017 Vavr, http://vavr.io
```
same here: `2014-2019`

cardil commented:
@alwins0n Awesome! I didn't expected that fast reaction! 

One thing I noticed is that constraint property path is slightly different that that of JVM's List or Map. I don't know is that a problem really?!

* Map's key is invalid:
  * Vavr's: `map<K>[bad].<map key>` -- notice a `bad` is a value of key that is invalid
  * Jvm's: `map<K>[bad].<map key>` -- the same as in Vavr's map.
* Map's value is invalid:
  * Vavr's: `map<V>[].<map value>` -- notice a `""` is a value that's invalid (is blank, but shouldn't)
  * Jvm's: `map[b].<map value>` -- **here Jvm Map's invalid value is reported** with key for which the value is invalid.
* Vavr's Seq: `list[].<iterable element>` -- there is no position on list
* Jvm's list: `list[2].<list element>` -- **there is exact position** of an error

alwins0n commented:
Nice observation. I guess for all Vavr Collections where a `get` with an index is possible ( `Seq`s and subtypes therof) a custom extractor with the exact position makes sense.

I'll try it out asap

alwins0n commented:
> 
>     * Vavr's Seq: `list[].<iterable element>` -- there is no position on list
> 
>     * Jvm's list: `list[2].<list element>` -- **there is exact position** of an error

I just realized a small problem with my mentioned approach. There is a severe performance penalty for extracting values from a vavr`List` if you access the elements by index, since the implementation is O(n) for array like access.

so the complexity of returning the exact index of the violation is O(n^2) since the list is traveresed in a `forEach` with index which is used to retrieve the element.

I will fix my implementation to always retrieve the head element and save the current index in some variable in an for loop instead of using `forEach`.

alwins0n commented:
@cardil have a look at the latest changes. All the topics mentioned in this PR should be covered now.

cardil commented:
Awesome. I'll review it tomorrow.

cardil commented on README.md:
```diff
@@ -5,11 +5,14 @@
 - `@NotEmpty` for vavr's `Value<T>`
-- All available validations can be applied to nested Tuple Values. see example below
+- All available validations can be applied to nested Tuple Values. See example below
+- All available validations can be applied to vavr's `Map<K, V>`s and `Traversable<T>`s
```
I've added `Traversable` here. Is it supported? Or it should be `Seq` here?

alwins0n replied:
Good question, I thought it was self explanatory but it is not. I will think about it.
Essentially it works on all vavr types since the inherit from `Iterable`.

I think I will add some more examples

cardil commented on src/test/java/io/vavr/beanvalidation2/valueextraction/MapValueExtractorTest.java:
```diff
@@ -0,0 +1,147 @@
+        bean.put("b", "");
+        validateAndAssertSingleViolation(
+                bean, NotBlank.class, "b",
+                "<V>", "value"
```
Java's Map produces output here: `map[b]<map value>`, but Vavr's adds additional `<V>` so it became: `map<V>[b]<map value>`. I don't know is that an issue or not.

I'm leaning to the side it doesn't matter.

alwins0n replied:
tbh, since we are using the interface correctly I think (of `ValueExtractor`) what the actual implementation (hibernate-validator) does with it - we should not care. that means we should not do string comparisons but rather query the `javax.validation.Path` via .getPropertyPath and not inspect the generated human readable string.

this way the tests could also be run against another implementation (non hibernate-validation). Can you rewrite it?

cardil replied:
I have mixed opinion on that. I agree, that we shouldn't rely on text representation of `javax.validation.Path`.

But this interface was designed for Java EE world when validation was done in back-end and validated values ware also presented in back-end. In that way one could rely on Java object representation of Path to properly notify on error, (in JSF it was automatic).

Nowadays validation results are mostly propagated to front-end client with some transport (mostly REST). Those transports are usually text based. Because of that people actually rely on output of `toString()` for Path as this is their only way of "serializing" validation violation across network.

Because of that, I've added those tests - to check the text representation that in fact matters for most JS/Java applications out there.

What do you think?

alwins0n replied:
Excellent question. But in my opinion this is not the responsibility of the validation provider to generate parseable path output here, rather the responsibility of the incorporating (REST-) framework such as spring, etc.

If bean validation is built-in those frameworks they are responsible for calling the provider and responding with a 400 providing additional information in a documented format. If not and you call the validation provider yourself, YOU are responsible to generate a 400 with information that you get back from the java API and make it "client-readable"

Bean validation should not care about REST, Change my view :)

cardil replied:
You're right that isn't responsibility of bean validation.

Unfortunately nobody (framework's authors) wants to gets their hands dirty and serialization of constraint violation is left to the users. I don't recall any framework actually do something better then call a `toString()` method on Path. Also, most StackOverflow articles advice just that (for ex: https://stackoverflow.com/a/44308282/844449).

Let's leave it as it is. It's good enough. So, I propose to leave those tests, as a reminder of that. They give only a little bit different output - I think that's okey.

alwins0n replied:
If the output were strictly json path or something I would say yeah, let's go for it, but it is not and vendor dependent. Maybe a good idea for a library on its own (bv-jsonpath)

I don't want the tests to break on library upgrade or implementation change. So I think we should cut dependency on tostring entirely

cardil replied:
Sure. I will rewrite those asserts to rely on on BV standard, so it will be implementation independent.

BTW. bv-jsonpath is really good idea!

cardil commented on src/test/java/io/vavr/beanvalidation2/valueextraction/SeqValueExtractorTest.java:
```diff
@@ -0,0 +1,118 @@
+    public void havingEmptyValueShouldNotValidate() {
+        TestBean bean = new TestBean();
+        bean.add("");
+        validateAndAssertSingleViolation(bean, 2, "seq");
```
Does it should be `"seq"` or different depending on exact type: `"list", "seq", "set", "stream", "queue"`. Is `Traversable` also supported?

alwins0n replied:
it is `"list"` for java in the hibernate validator implementation ( see https://github.com/hibernate/hibernate-validator/blob/master/engine/src/main/java/org/hibernate/validator/internal/engine/path/NodeImpl.java ) which is the interface for all list like types so I figured "`seq`" was fine

alwins0n replied:
`Traversable` is essentially an `Iterable` type and in scala 2.13 they actually renamed it. I proposed the same for `vavr` 1.0.0

So to stay with "`<iterable element>`" as catchall for all others is fine I think. ( see again https://github.com/hibernate/hibernate-validator/blob/master/engine/src/main/java/org/hibernate/validator/internal/engine/path/NodeImpl.java )

alwins0n replied:
On the other hand: hibernate-validator is an implementation of bean validation of which there could be multiple. `NodeImpl` is from hibernate-validator and we should probably not put too much weight on how they do things over there.

In the end there are 3 sensible feedbacks for generic vavr types (implementing `Value<T>`:
- an error occured at a certain index of the vavr type (Seqs
- an error occured at a certain key of the vavr type (Maps)
- an error occured SOMEWHERE in the generic type (Iterables, could be a Set or an Either,....)

Beanvalidations `ValueReceiver` distinguishes between 4 types:
- value
- iterableValue
- indexedValue
- keyedValue

Maybe we should do the same: so here is what i came up with
- indexedValue with feedback "`<index element>`" for Seq and below
- keyedValue with feedback "`<map key>`" or "`<map value>`" for Map and MultiMap (!!!)
- iterableValue with feedabck "`<iterable element>`" for all Traversables
- value with feedback "`<nested element>`" or "`<value element>`" for all other `Value<T>`

actually Multimap is a bit of a challenge here. since it combines keyed with indexed/iterable somewhat so I am open for suggestions (or introduce an own feedback type)

alwins0n replied:
shame that `Value` will be deprecated soon, in this context I really liked the separation as outlined above.

but we should bear in mind this issue: https://github.com/vavr-io/vavr/issues/2447

cardil replied:
I vote for: 

 * `<iterable element>` for Traversables
 * `<sequence element>` for Seq and below
 * `<map key>` and `<map value>` for Map and Multimap,
 * `<element>` for Values

alwins0n replied:
I almost agree on your proposal. 

Not 100% sure about sequence though. The API says `indexedValue` because it is the only thing that matters. A collection whose elements can be retrieved via index, and whose errors can be debugged by knowing the index.

So we dont care about it being a sequence or not, just "a type that you can query via index" as opposed to iterable or map.
This type being called `Seq` in vavr is merely a "coincidence", imho.

On the other hand, hibernate validator also calls it "list element".

Maybe "`<indexed element>`" or am I thinking too abstract here? 

What do you think?

cardil replied:
I think that the [Vavr's Seq](https://static.javadoc.io/io.vavr/vavr/0.10.0/io/vavr/collection/Seq.html) is named directly after [Scala's Seq](https://www.scala-lang.org/api/2.12.3/scala/collection/Seq.html). Both of them describe themselves as sequential, not only indexed.

Scala's: 
> sequences always have a defined order of elements

Vavr's:
> immutable sequential data structures

I opt for name `<sequence element>` as the type is called Seq, and describe it self as sequential. Indexed isn't mentioned anywhere. Using the same name as type and Javadoc used is more straight forward and easier to understand for users. 

In fact in both in Scala and in Vavr there are: `IndexedSeq` and `LinearSeq` both inheriting from `Seq`. So, `<sequence element>` seams a good description to me.

alwins0n replied:
You convinced me. Let's go for sequence element here

cardil commented:
I've added support for `Multimap`, as it was missing. `Either` as well. I've introduced changes discussed in code review process. Exception to those changes was `Either` support, as it is needed to differentiate between right and left element.

On top that, there are some quirks:

 * Validating a `Either` element is tricky as extractors try to extract both values from it, while validating. Users might expect that given a right `Either` only right value will be validated, not both.
 * Option element is validated but element is named `<iterable element>` not `<element>`

Kindly, please review @alwins0n

cardil commented on .travis.yml:
```diff
@@ -1,5 +1,6 @@
 language: java
-
+sudo: false
+dist: trusty
```
I had to add such changes because it stopped building: https://travis-ci.org/vavr-io/vavr-beanvalidation2/builds/566001535

alwins0n replied:
actually we should add openjdk11. I will do that

alwins0n replied:
sigh, travis build fails due to jacoco. I will probably have to talk to daniel about this. will do later

cardil commented on src/test/java/io/vavr/beanvalidation2/valueextraction/OptionTest.java:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	@Test
+	@Ignore("On line 79 value is <iterable element> instead of <element>")
```
This should be fixed in some way

alwins0n replied:
see discussion below (comment on 2019-08-03)

alwins0n commented:
> I've added support for `Multimap`, as it was missing. `Either` as well. I've introduced changes discussed in code review process. Exception to those changes was `Either` support, as it is needed to differentiate between right and left element.
> 
> On top that, there are some quirks:
> 
> * Validating a `Either` element is tricky as extractors try to extract both values from it, while validating. Users might expect that given a right `Either` only right value will be validated, not both.
> * Option element is validated but element is named `<iterable element>` not `<element>`

thank you for your work. to respond to your remarks:

- I find `Either` to be not troublesome. The left is extracted as null value and since all Validators are capable of handling null properly I see no problem. If you want to ensure a right you would to for `Either<@NotNull Integer, String>` or even annotating the whole either with `@NotEmpty` since it is iterable and iterates over the "right" value. I think the latter is the way to go!
- I find the message for `Option` also not too problematic since you can view Option to be a collection of size 0 or 1. BUT: the problem holds for `Try` as well. The question is: "Does it make sense to report <iterable element> in case of validation of Try?". Imho the usecase for validating a `Try` is very rare. Maybe document it as a known limitation/issue?
- `MultiMap` is a lot more complicated I think (if seq-backed). Yes, it does help to know at which key a validation error occured, but then what? under that key there could be a million (sequential?) values! Problably also a "known limitation" since the BV API does not allow for key AND index. Document?

cardil commented:
> * I find `Either` to be not troublesome. The left is extracted as null value and since all Validators are capable of handling null properly I see no problem. If you want to ensure a right you would to for `Either<@NotNull Integer, String>` or even annotating the whole either with `@NotEmpty` since it is iterable and iterates over the "right" value. I think the latter is the way to go!

Agreed. It's might be troublesome for users. They can't use builtin validators like `@NotBlank`. If do it's always fail as one of the values will be `null`, by design of `Either`. It can be validated, but requires a specialized validators, that allow a null values.

> * I find the message for `Option` also not too problematic since you can view Option to be a collection of size 0 or 1. BUT: the problem holds for `Try` as well. The question is: "Does it make sense to report  in case of validation of Try?". Imho the usecase for validating a `Try` is very rare. Maybe document it as a known limitation/issue?

Agreed. I will remove an ignore. Should you write a documentation on known limitations?

> * `MultiMap` is a lot more complicated I think (if seq-backed). Yes, it does help to know at which key a validation error occured, but then what? under that key there could be a million (sequential?) values! Problably also a "known limitation" since the BV API does not allow for key AND index. Document?

I find `Multimap` as not problematic. Multimaps has this contract that there are several values that all are switchable are everyone of them is assigned to same key. It might be useful to show exact place, but it isn't required as in my opinion, key index is enough in this case.

cardil commented:
This PR should close #6  due to 0e99fce commit.

alwins0n commented:
I will reread the PR once more soon, but I think we are rdy to merge

alwins0n commented:
updated the doc, I am ready to go. @cardil lets merge?

cardil commented:
:ok_hand: LGTM
